src/sentry/utils/snuba.py

```diff
@@ -903,6 +906,135 @@ def nest_groups(data, groups, aggregate_cols):
+    if col in DATASETS[dataset]:
+        return DATASETS[dataset][col]
+    mapping = DATASETS[dataset]
+    if col in mapping.values():
+        return col
+    return u"tags[{}]".format(col)
+
+
+def constrain_condition_to_dataset(cond, dataset):
+    """
+    When conditions have been parsed by the api.event_search module
+    we can end up with conditions that are not valid on the current dataset
+    due to how ap.event_search checks for valid field names without
+    being aware of the dataset.
+
+    We have the dataset context here, so we need to re-scope conditions to the
+    current dataset.
+    """
+    if isinstance(cond, (list, tuple)) and len(cond):
+        if isinstance(cond[0], (list, tuple)):
+            # Nested condition or function expressions
+            cond = [constrain_condition_to_dataset(c, dataset) for c in cond]
+        elif len(cond) == 3:
+            # map column name
+            name = constrain_column_to_dataset(cond[0], dataset)
+            if name is None:
+                return None
+            cond[0] = name
+            # Reformat 32 byte uuids to 36 byte variants.
+            # The transactions dataset requires properly formatted uuid values.
+            # But the rest of sentry isn't aware of that requirement.
+            if dataset == TRANSACTIONS and name == "event_id" and len(cond[2]) == 32:
+                cond[2] = six.text_type(uuid.UUID(cond[2]))
+        elif len(cond) == 2 and cond[0] == "has":
+            # first function argument is the column if function is "has"
+            cond[1][0] = constrain_column_to_dataset(cond[1][0], dataset)
+    return cond
+
+
+def dataset_query(
+    start=None,
+    end=None,
+    groupby=None,
+    conditions=None,
+    filter_keys=None,
+    aggregations=None,
+    selected_columns=None,
+    arrayjoin=None,
+    having=None,
+    dataset=None,
+    orderby=None,
+    **kwargs
+):
+    """
+    Wrapper around raw_query that selects the dataset based on the
+    selected_columns, conditions and groupby parameters.
+    Useful for taking arbitrary end user queries and searching
+    either error or transaction events.
+
+    This function will also re-alias columns to match the selected dataset
+    """
+    if dataset is None:
+        dataset = detect_dataset(
+            dict(
+                dataset=dataset,
+                aggregations=aggregations,
+                conditions=conditions,
+                selected_columns=selected_columns,
+                groupby=groupby,
+            )
+        )
+
+    def remove_none(values):
+        return list(filter(lambda x: x is not None, values))
+
+    derived_columns = []
+    if selected_columns:
+        for (i, col) in enumerate(selected_columns):
+            if isinstance(col, list):
+                derived_columns.append(col[2])
+            else:
+                selected_columns[i] = constrain_column_to_dataset(col, dataset)
+        selected_columns = remove_none(selected_columns)
```

Comment:
There is a bit of duplication between this function and `transform_aliases_and_query()`. I plan to add several integration tests and then refactor the duplication out where possible.